src/mergeDeep.ts

```diff
@@ -0,0 +1,95 @@
+import { purry } from './purry';
+import { MergeDeep } from './type-fest/merge-deep';
+
+/**
+ * Merges the `source` object into the `destination` object, while recursively merging any matching nested objects. When both objects have a value for a key and either value is not an object, the value from `source` would overwrite the value from `destination` (similar to a simple spread `{ ...destination, ...source }`).
+ *
+ * @param destination - The object to merge into. In general, this object would have it's values overridden.
+ * @param source - The object to merge from. In general, shared keys would be taken from this object.
+ * @returns - The merged object.
+ * @signature
+ *    R.mergeDeep(destination, source)
+ * @example
+ *    R.mergeDeep({ foo: 'bar', x: 1 }, { foo: 'baz', y: 2 }) // => { foo: 'baz', x: 1, y: 2 }
+ * @dataFirst
+ * @category Object
+ */
+export function mergeDeep<
+  Destination extends Record<string, unknown>,
+  Source extends Record<string, unknown>,
+>(destination: Destination, source: Source): MergeDeep<Destination, Source>;
+
+/**
+ * Merges the `source` object into the `destination` object, while recursively merging any matching nested objects. When both objects have a value for a key and either value is not an object, the value from `source` would overwrite the value from `destination` (similar to a simple spread `{ ...destination, ...source }`).
```

Comment:
**nit:** the "either value is not an object" phrasing is a bit awkward?

**suggestion:**

```suggestion
 * Merges the `source` object into the `destination` object. When both objects have a value for a key:
 * - if both values are objects, they are recursively merged;
 * - otherwise, the value from `source` will be used (similar to a simple spread `{ ...destination, ...source }`).
```